src/wpas/wpa_controller.c

```diff
@@ -784,6 +784,7 @@ __wpa_controller_send_commandf(struct wpa_controller *ctrl, const char *name, ch
         zlog_error_if(ctrl->interface, "failed to send %s command on ctrl interface (%d)", name, ret);
         return ret;
     }
+    reply[*reply_length > WPA_MAX_MSG_SIZE? WPA_MAX_MSG_SIZE : *reply_length] = '\0';
```

Comment:
This will truncate all replies > `WPA_MAX_MSG_SIZE`, even if the caller provided a large enough buffer. Also, this truncates the reply so would have to return `-EOVERFLOW` to indicate this. The caller could just check `reply_size` to see if it was larger than the original, but then that would force them to introduce a new variable to cache the original, which is duplicated effort on the part of each caller. Also, the return value is meant to be the single source-of-truth regarding whether the function succeeded or failed (contrasting with other styles, like setting `errno`).

Something like this is pretty straight-forward:
```C
size_t reply_length_actual = reply_length;
ret = wpa_ctrl_request(ctrl->command, cmd, cmd_length, reply, &reply_length_actual, NULL);
if (ret < 0) {
   // ...
   return ret;
} else if (reply_length_actual < reply_length) {
  reply[reply_length_actual] = '\0';
} else {
  ret = -EOVERFLOW;
  reply[*reply_length - 1] = '\0';
}

zlog_debug_if(...);

return ret;
```

Using a separate variable for the real reply length costs almost nothing and provides callers ultimate flexibility to choose how to handle misbehaving remote processes.

Reply:
Oh, now I understand the role of reply_length. I had always thought it was just a dummy variable used to store the size of the reply of wpa_ctrl_request.

Reply:
Is there a reason that reply_length could be less than 4097 when we first enter __wpa_controller_send_commandf? Otherwise would it be valid to restructure the function so that it assumes the reply buffer is length 4097?

Reply:
When first entering `__wpa_controller_send_commandf`, `*reply_length` could be anything. We don't actually care how big it is, and have no real reason to limit it; we just marshal the reply from the lib to the buffer in a way that's safe and makes no assumptions about outside callers.

If there was a reason to make assumptions, it would be _far_ safer to actually enforce the assumption so that it becomes a pre-condition. But without enforcement, the code could change at anytime and the assumption becomes invalid; a bug results.

Especially so in this case: the proposed solution isn't suer complicated or complex: it just makes sure we don't overflow the provided buffer.